src/hotspot/os/aix/porting_aix.cpp

```diff
@@ -891,3 +897,267 @@ bool AixMisc::query_stack_bounds_for_current_thread(stackbounds_t* out) {
+    combined.print("%s/%s", token, path2);
+    if ((ret = (0 == stat64x(combined.base(), stat))))
+      break;
+  }
+
+  os::free (libpath);
+  os::free (path2);
+  return ret;
+}
+
+// specific AIX versions for ::dlopen() and ::dlclose(), which handles the struct g_handletable
+// filled by os::dll_load(). This way we mimic dl handle equality for a library
+// opened a second time, as it is implemented on other platforms.
+void* Aix_dlopen(const char* filename, int Flags, char *ebuf, int ebuflen) {
+  void* result;
+  struct stat64x libstat;
+
+  if (false == search_file_in_LIBPATH(filename, &libstat)) {
+    // file with filename does not exist
+  #ifdef ASSERT
+    result = ::dlopen(filename, Flags);
+    assert(result == nullptr, "dll_load: Could not stat() file %s, but dlopen() worked; Have to improve stat()", filename);
+  #endif
+  }
+  else {
+    int i = 0;
+    pthread_mutex_lock(&g_handletable_mutex);
+    // check if library belonging to filename is already loaded.
+    // If yes use stored handle from previous ::dlopen() and increase refcount
+    for (i = 0; i < g_handletable_used; i++) {
+      if (g_handletable[i].handle &&
+          g_handletable[i].inode == libstat.st_ino &&
+          g_handletable[i].devid == libstat.st_dev) {
+        g_handletable[i].refcount++;
+        result = g_handletable[i].handle;
+        break;
+      }
+    }
+    if (i == g_handletable_used) {
+      // library not yet loaded. Check if there is space left in array
+      // to store new ::dlopen() handle
+      if (g_handletable_used == max_handletable) {
+        // Array is already full. No place for new handle. Cry and give up.
+        pthread_mutex_unlock(&g_handletable_mutex);
+        if (ebuf != nullptr && ebuflen > 0) {
+          ::strncpy(ebuf, "dlopen: too many libraries loaded", ebuflen - 1);
+        }
+        assert(false, "max_handletable reached");
+        return nullptr;
+      }
+      // Library not yet loaded; load it, then store its handle in handle table
+      result = ::dlopen(filename, Flags);
+      if (result != nullptr) {
+        g_handletable_used++;
+        g_handletable[i].handle = result;
+        g_handletable[i].inode = libstat.st_ino;
+        g_handletable[i].devid = libstat.st_dev;
+        g_handletable[i].refcount = 1;
+      }
+      else {
+        // error analysis when dlopen fails
+        const char* error_report = ::dlerror();
+        if (error_report == nullptr) {
+          error_report = "dlerror returned no error description";
+        }
+        if (ebuf != nullptr && ebuflen > 0) {
+          snprintf(ebuf, ebuflen - 1, "%s", error_report);
+        }
+      }
+    }
+    pthread_mutex_unlock(&g_handletable_mutex);
+  }
+  return result;
+}
+
+bool os::pd_dll_unload(void* libhandle, char* ebuf, int ebuflen) {
+  int i = 0;
+  bool res = false;
+
+  if (ebuf && ebuflen > 0) {
+    ebuf[0] = '\0';
+    ebuf[ebuflen - 1] = '\0';
+  }
+
+  pthread_mutex_lock(&g_handletable_mutex);
```

Comment:
You can make your life a lot easier by defining an RAII object at the start of the file:
```
struct TableLocker {
  TableLocker() { pthread_mutex_lock(&g_handletable_mutex); }
  ~TableLocker() { pthread_mutex_unlock(&g_handletable_mutex); }
};
```
and just place this at the beginning of your two functions
```
TableLocker lock:
...
```

no need to manually unlock then, with the danger of missing a return.

Reply:
Great, thank you. This was one of the things I thought about, but was not sure, because I did not fully understood the MutexLocker class and the difference between Monitor and Mutex.

Reply:
In hindsight, pthread mutex is the better choice anyway: it avoids dependencies to the VM lifecycle (VM mutexes are only available after VM initialization, so we could not call dlopen() before that).